File: force/rust/src/parser.rs
use crate::defs::*;
use crate::lexer::{lexer, Token};
use logos::Span;
use regex::Regex;
use std::collections::HashMap;
use std::sync::Arc;

pub struct Parser {
    tokens: Vec<(Token, Span)>,
    count: usize,
    cur: Token,
    source: String,
}

enum Choice {
    Category(String),
    Family(String),
}

impl Parser {
    pub fn new(source: &str) -> Parser {
        let tokens = lexer(source);
        return Parser {
            count: 0,
            cur: tokens[0].0.clone(),
            tokens: tokens,
            source: source.to_owned(),
        };
    }
    fn advance(&mut self) {
        self.count += 1;
        self.cur = self.tokens[self.count].clone().0
    }
    fn eat(&mut self, expect: Token) -> ForceResult<()> {
        if self.cur == expect {
            self.advance();
            Ok(())
        } else {
            Err(ForceError::NonExpect {
                expect,
                fact: self.cur.clone(),
            })
        }
    }
    fn get_integer(&mut self) -> ForceResult<usize> {
        let ret = if let Token::Integer(n) = &self.cur {
            Ok(*n)
        } else {
            Err(ForceError::NonExpect {
                // TODO: 提高拋錯的可讀性
                expect: Token::Integer(123456789),
                fact: self.cur.clone(),
            })
        };
        if let Ok(_) = ret {
            self.advance();
        }
        ret
    }
    fn get_identifier(&mut self) -> ForceResult<String> {
        let ret = if let Token::Identifier(id) = &self.cur {
            Ok(id.clone())
        } else {
            Err(ForceError::NonExpect {
                expect: Token::Identifier("某個識別子".to_owned()),
                fact: self.cur.clone(),
            })
        };
        if let Ok(_) = ret {
            self.advance();
        }
        ret
    }
    fn parse_choice(&mut self) -> ForceResult<Choice> {
        match self.cur {
            Token::At => {
                self.advance();
                let name = self.get_identifier()?;
                Ok(Choice::Family(name))
            }
            _ => {
                let name = self.get_identifier()?;
                Ok(Choice::Category(name))
            }
        }
    }
    fn parse_choices(&mut self) -> ForceResult<Bondee> {
        let mut category = vec![];
        let mut family = vec![];
        match self.parse_choice()? {
            Choice::Category(name) => category.push(name),
            Choice::Family(name) => family.push(name),
        }
        loop {
            match self.cur {
                Token::RightSquareBracket => {
                    break;
                }
                Token::Comma => {
                    self.advance();
                    match self.parse_choice()? {
                        Choice::Category(name) => category.push(name),
                        Choice::Family(name) => family.push(name),
                    }
                }
                _ => {
                    return Err(ForceError::NoMeet {
                        expect: ", 或 ]".to_owned(),
                        fact: self.cur.clone(),
                    });
                }
            }
        }
        Ok(Bondee::Choices { category, family })
    }
    fn parse_bondee(&mut self) -> ForceResult<Bondee> {
        self.eat(Token::LeftSquareBracket)?;
        match self.cur.clone() {
            Token::Star => {
                self.advance();
                self.eat(Token::RightSquareBracket)?;
                Ok(Bondee::All)
            }
            Token::At | Token::Identifier(_) => {
                let choices = self.parse_choices()?;
                self.eat(Token::RightSquareBracket)?;
                Ok(choices)
            }
            _ => Err(ForceError::NoMeet {
                expect: "* 或識別子".to_owned(),
                fact: self.cur.clone(),
            }),
        }
    }
    fn parse_datatype(&mut self) -> ForceResult<BasicDataType> {
        match self.cur {
            Token::Number => {
                self.advance();
                Ok(BasicDataType::Number)
            }
            Token::OneLine => {
                self.advance();
                Ok(BasicDataType::OneLine)
            }
            Token::Text => {
                self.advance();
                match self.cur.clone() {
                    Token::Regex(s) => {
                        self.advance();
                        let regex = Regex::new(&format!("(?s){}", s))
                            .map_err(|_e| ForceError::InvalidRegex { regex: s })?;
                        Ok(BasicDataType::Text(Some(regex)))
                    }
                    _ => Ok(BasicDataType::Text(None)),
                }
            }
            Token::Bond => {
                self.advance();
                let bondee = self.parse_bondee()?;
                Ok(BasicDataType::Bond(bondee))
            }
            _ => Err(ForceError::NoMeet {
                expect: "型別".to_owned(),
                fact: self.cur.clone(),
            }),
        }
    }
    pub fn parse_family(&mut self) -> ForceResult<Vec<String>> {
        match self.cur.clone() {
            Token::At => {
                self.advance();
                self.eat(Token::LeftSquareBracket)?;
                // 不允許分類族為空
                let name = self.get_identifier()?;
                let mut family = vec![name];
                loop {
                    match self.cur {
                        Token::RightSquareBracket => {
                            self.advance();
                            break;
                        }
                        _ => {
                            self.eat(Token::Comma)?;
                            let name = self.get_identifier()?;
                            family.push(name);
                        }
                    }
                }
                Ok(family)
            }
            _ => Ok(vec![]),
        }
    }
    pub fn parse_field(&mut self) -> ForceResult<Field> {
        let basic_datatype = self.parse_datatype()?;
        let name = self.get_identifier()?;
        let datatype = match self.cur {
            Token::QuestionMark => {
                self.advance();
                DataType::Optional(basic_datatype)
            }
            Token::LeftSquareBracket => {
                self.advance();
                let min = self.get_integer()?;
                self.eat(Token::Tilde)?;
                let max = self.get_integer()?;
                self.eat(Token::RightSquareBracket)?;
                DataType::Array {
                    t: basic_datatype,
                    min,
                    max,
                }
            }
            _ => DataType::Single(basic_datatype),
        };
        Ok(Field { datatype, name })
    }
    pub fn parse_category(&mut self) -> ForceResult<Category> {
        let start = self.tokens[self.count].1.start;
        // 讀取分類名稱
        let name = self.get_identifier()?;

        // 讀取分類族
        let family = self.parse_family()?;

        // 讀取各欄位資訊
        let mut fields = Vec::new();
        self.eat(Token::LeftCurlyBrace)?;
        loop {
            if let Token::RightCurlyBrace = self.cur {
                break;
            } else {
                fields.push(self.parse_field()?);
            }
        }
        let end = self.tokens[self.count].1.end;
        self.eat(Token::RightCurlyBrace)?;
        Ok(Category {
            name,
            fields,
            family,
            source: self.source[start..end].to_string(),
        })
    }
    fn parse_categories(&mut self) -> ForceResult<Categories> {
        let mut categories = HashMap::new();
        loop {
            if let Token::End = self.cur {
                break;
            } else {
                let category = Arc::new(self.parse_category()?);
                categories.insert(category.name.clone(), category);
            }
        }
        return Ok(categories);
    }
    pub fn parse(&mut self) -> ForceResult<Force> {
        let categories = self.parse_categories()?;

        // 建造分類族雜湊表
        let mut families: HashMap<String, Vec<String>> = HashMap::new();

        for (_key, category) in &categories {
            for family in &category.family {
                match families.get_mut(family) {
                    Some(f) => {
                        f.push(category.name.clone());
                    }
                    None => {
                        families.insert(family.clone(), vec![category.name.clone()]);
                    }
                }
            }
        }

        // 檢驗鍵結指向的分類跟分類族是否存在
        let mut not_found_categories = Vec::new();
        let mut not_found_families = Vec::new();

        for (_key, category) in &categories {
            for field in &category.fields {
                match field.datatype.basic_type() {
                    BasicDataType::Bond(bondee) => {
                        if let Bondee::Choices {
                            family: family_choices,
                            category: category_choices,
                        } = bondee
                        {
                            for c in category_choices {
                                if categories.get(c).is_none() {
                                    not_found_categories.push(c.clone());
                                }
                            }
                            for f in family_choices {
                                if families.get(f).is_none() {
                                    not_found_families.push(f.clone());
                                }
                            }
                        }
                    }
                    _ => {}
                }
            }
        }

        if not_found_categories.len() > 0 || not_found_families.len() > 0 {
            return Err(ForceError::InvalidBond {
                not_found_categories,
                not_found_families,
            });
        }

        return Ok(Force {
            categories,
            families,
        });
    }
}

pub fn parse(source: &str) -> ForceResult<Force> {
    Parser::new(source).parse()
}

pub fn parse_category(source: &str) -> ForceResult<Category> {
    Parser::new(source).parse_category()
}

#[cfg(test)]
mod test {
    use super::*;
    #[test]
    fn test_simple_category() -> ForceResult<()> {
        let source = "新聞 @[轉載, 外部] {單行 記者 單行 網址}";

        let force = parse(source)?;
        assert_eq!(force.categories.len(), 1);

        let ans = Category {
            name: "新聞".to_owned(),
            fields: vec![
                Field {
                    datatype: BasicDataType::OneLine.into(),
                    name: "記者".to_owned(),
                },
                Field {
                    datatype: BasicDataType::OneLine.into(),
                    name: "網址".to_owned(),
                },
            ],
            family: vec!["轉載".to_owned(), "外部".to_owned()],
            source: source.to_owned(),
        };
        assert_eq!(**(force.categories.get("新聞").unwrap()), ans);
        assert_eq!(parse_category(source).unwrap(), ans);

        assert_eq!(force.families.len(), 2);
        assert_eq!(force.families.get("轉載").unwrap().len(), 1);
        assert_eq!(
            force
                .families
                .get("轉載")
                .unwrap()
                .contains(&"新聞".to_string()),
            true
        );
        assert_eq!(force.families.get("外部").unwrap().len(), 1);
        assert_eq!(
            force
                .families
                .get("外部")
                .unwrap()
                .contains(&"新聞".to_string()),
            true
        );

        Ok(())
    }
    #[test]
    fn test_regex() -> ForceResult<()> {
        let source = "作文比賽 {文本/我的志願是.+/ 文章}";

        let force = parse(source)?;
        assert_eq!(force.categories.len(), 1);

        let ans = &Category {
            name: "作文比賽".to_owned(),
            fields: vec![Field {
                datatype: BasicDataType::Text(Some(Regex::new("(?s)我的志願是.+").unwrap())).into(),
                name: "文章".to_owned(),
            }],
            family: vec![],
            source: source.to_owned(),
        };
        assert_eq!(&parse_category(source).unwrap(), ans);
        Ok(())
    }
    #[test]
    fn test_choices() -> ForceResult<()> {
        let source = "留言 { 鍵結[@批踢踢文章, @狄卡文章, 新聞] 原文 }";

        let ans = &Category {
            name: "留言".to_owned(),
            fields: vec![Field {
                datatype: BasicDataType::Bond(Bondee::Choices {
                    category: vec!["新聞".to_owned()],
                    family: vec!["批踢踢文章".to_owned(), "狄卡文章".to_owned()],
                })
                .into(),
                name: "原文".to_owned(),
            }],
            family: vec![],
            source: source.to_owned(),
        };
        assert_eq!(&parse_category(source).unwrap(), ans);
        Ok(())
    }
    #[test]
    fn test_family() -> ForceResult<()> {
        let source = "
            留言 { 鍵結[@批踢踢文章, @狄卡文章, 新聞] 原文 }
            新聞 {}
            八卦 @[批踢踢文章] {}
            政黑 @[批踢踢文章] {}
            有趣 @[狄卡文章] {}
            ";
        let force = parse(source)?;

        assert_eq!(force.families.get("批踢踢文章").unwrap().len(), 2);
        assert_eq!(
            force
                .families
                .get("批踢踢文章")
                .unwrap()
                .contains(&"八卦".to_string()),
            true
        );
        assert_eq!(
            force
                .families
                .get("批踢踢文章")
                .unwrap()
                .contains(&"政黑".to_string()),
            true
        );

        assert_eq!(force.families.get("狄卡文章").unwrap().len(), 1);
        assert_eq!(
            force
                .families
                .get("狄卡文章")
                .unwrap()
                .contains(&"有趣".to_string()),
            true
        );
        Ok(())
    }
}

File: force/typescript/src/parser.ts
import { lexer } from './lexer';
import * as moo from 'moo';
import { Bondee, BasicDataType, DataType, Category, Categories, Force, Field } from './defs';

function non_expect(expect: string, fact: moo.Token): Error {
	return new Error(`預期 ${expect} ，但得到 ${JSON.stringify(fact)}`);
}

class SemanticError extends Error {
	constructor(message: string) {
		super(message);
		this.name = 'SemanticError';
	}
}

function invalid_bond(categories: string[], families: string[]): SemanticError {
	let msg;
	if (families.length == 0) {
		msg = `不存在分類 [ ${categories} ]`;
	} else if (categories.length == 0) {
		msg = `不存在分類族 [ ${families.map(f => '@' + f)} ]`;
	} else {
		msg = `不存在分類 [ ${categories} ]
		以及分類族 [ ${families.map(f => '@' + f)} ]
`;
	}
	return new SemanticError(msg);
}

type Choice = {
	kind: 'category',
	name: string,
} | {
	kind: 'family',
	name: string,
};

export class Parser {
	source: string;
	tokens: moo.Token[];
	count: number;
	constructor(s: string) {
		this.source = s;
		this.tokens = [];
		this.count = 0;
		lexer.reset(this.source);
		while (true) {
			const token = lexer.next();
			if (token == undefined) {
				break;
			} else if (token.type == 'new_line' || token.type == 'whitespace') {
				continue;
			}
			this.tokens.push(token);
		}
	}
	is_end(): boolean {
		return this.tokens.length == this.count;
	}
	cur(): moo.Token {
		if (this.count == this.tokens.length) {
			throw new Error('已到 tokens 結尾');
		}
		return this.tokens[this.count];
	}
	advance(): void {
		this.count = Math.min(this.count + 1, this.tokens.length);
	}
	eat(expect: string): void {
		if (this.cur().type == expect) {
			this.advance();
		} else {
			throw non_expect(expect, this.cur());
		}
	}
	get_identifier(): string {
		if (this.cur().type == 'identifier') {
			const ret = this.cur().value;
			this.advance();
			return ret;
		} else {
			throw non_expect('identifier', this.cur());
		}
	}
	get_integer(): number {
		if (this.cur().type == 'integer') {
			const ret = this.cur().value;
			this.advance();
			return parseInt(ret);
		} else {
			throw non_expect('integer', this.cur());
		}
	}
	parse_choice(): Choice {
		switch (this.cur().type) {
			case 'at': {
				this.advance();
				const name = this.get_identifier();
				return { kind: 'family', name };
			}
			default: {
				const name = this.get_identifier();
				return { kind: 'category', name };
			}
		}
	}
	parse_choices(): Bondee {
		const category: string[] = [];
		const family: string[] = [];
		const choice = this.parse_choice();
		if (choice.kind == 'category') {
			category.push(choice.name);
		} else if (choice.kind == 'family'){
			family.push(choice.name);
		}
		while (true) {
			if (this.cur().type == 'right_square_bracket') {
				break;
			} else if (this.cur().type == 'comma') {
				this.advance();
				const choice = this.parse_choice();
				if (choice.kind == 'category') {
					category.push(choice.name);
				} else if (choice.kind == 'family'){
					family.push(choice.name);
				}
			} else {
				throw non_expect(', 或 ]', this.cur());
			}
		}
		return {
			kind: 'choices',
			category,
			family
		};
	}
	parse_bondee(): Bondee {
		this.eat('left_square_bracket');
		switch (this.cur().type) {
			case 'star': {
				this.advance();
				this.eat('right_square_bracket');
				return { kind: 'all' };
			}
			case 'identifier':
			case 'at': {
				let choices = this.parse_choices();
				this.eat('right_square_bracket');
				return choices;
			}
			default: {
				throw non_expect('* 或識別子', this.cur());
			}
		}
	}
	parse_datatype(): BasicDataType {
		switch (this.cur().type) {
			case 'number': {
				this.advance();
				return {kind: 'number'};
			}
			case 'one_line': {
				this.advance();
				return {kind: 'one_line'};
			}
			case 'text': {
				this.advance();
				if (this.cur().type == 'regex') {
					const regex = this.cur().value;
					this.advance();
					return {
						kind: 'text',
						regex,
					};
				} else {
					return {
						kind: 'text',
						regex: undefined
					};
				}
			}
			case 'bond': {
				this.advance();
				const bondee = this.parse_bondee();
				return {
					kind: 'bond',
					bondee
				};
			}
			default: {
				throw non_expect('型別', this.cur());
			}
		}
	}
	parse_family(): string[] {
		switch (this.cur().type)  {
			case 'at': {
				this.advance();
				this.eat('left_square_bracket');
				const name = this.get_identifier();
				const family = [name];
				while (true) {
					if (this.cur().type == 'right_square_bracket') {
						this.advance();
						break;
					} else {
						this.eat('comma');
						const name = this.get_identifier();
						family.push(name);
					}
				}
				return family;
			}
			default: {
				return [];
			}
		}
	}
	parse_field(): Field {
		const basic_datatype = this.parse_datatype();
		const name = this.get_identifier();
		const datatype = ((): DataType => {
			if (this.cur().type == 'question_mark') {
				this.advance();
				return {kind: 'optional', t: basic_datatype};
			} else if (this.cur().type == 'left_square_bracket') {
				this.advance();
				const min = this.get_integer();
				this.eat('tilde');
				const max = this.get_integer();
				this.eat('right_square_bracket');
				return {
					kind: 'array',
					t: basic_datatype,
					min, max,
				};
			} else {
				return {kind: 'single', t: basic_datatype};
			}
		})();
		return { name, datatype };
	}
	parse_category(): Category {
		// 讀取分類名稱
		const name = this.get_identifier();

		// 讀取分類族
		const family = this.parse_family();

		// 讀取各欄位資訊
		const fields = [];
		this.eat('left_curly_brace');

		while (true) {
			if (this.cur().type == 'right_curly_brace') {
				break;
			} else {
				fields.push(this.parse_field());
			}
		}
		this.eat('right_curly_brace');
		return {
			name,
			family,
			fields
		};
	}
	parse_categories(): Categories {
		let categories = new Map<string, Category>();
		while (true) {
			if (this.is_end()) {
				break;
			} else {
				const category = this.parse_category();
				categories.set(category.name, category);
			}
		}
		return categories;
	}
	parse(): Force {
		const categories = this.parse_categories();

		// 建造分類族雜湊表
		const families = new Map<string, string[]>();

		for (let [_key, category] of categories) {
			for (let family of category.family) {
				let f = families.get(family);
				if (f == undefined) {
					families.set(family, [category.name]);
				} else {
					f.push(category.name);
				}
			}
		}

		// 檢驗鍵結指向的分類跟分類族是否存在
		let not_found_categories = new Set<string>();
		let not_found_families = new Set<string>();

		for (let [_key, category] of categories) {
			for (let field of category.fields) {
				if (field.datatype.t.kind == 'bond') {
					const bondee = field.datatype.t.bondee;
					if (bondee.kind == 'choices') {
						for (let c of bondee.category) {
							if (categories.get(c) == undefined) {
								not_found_categories.add(c);
							}
						}
						for (let f of bondee.family) {
							if (families.get(f) == undefined) {
								not_found_families.add(f);
							}
						}
					}
				}
			}
		}

		if (not_found_categories.size > 0 || not_found_families.size > 0) {
			throw invalid_bond([...not_found_categories], [...not_found_families]);
		}

		return {
			families,
			categories
		};
	}
}

export function parse(source: string): Force {
	const parser = new Parser(source);
	return parser.parse();
}

export function parse_category(source: string): Category {
	const parser = new Parser(source);
	return parser.parse_category();
}